### DISCUSSION TITLE
Airflow "This DAG isnt available in the webserver DagBag object "

### DISCUSSION THREAD

**Discussion by dotsinspace** [apache/airflow#30163] [General]
### Apache Airflow version

2.5.2

### What happened

NOTE: I am building application based on ReactFlow and Backend in Python using Airflow

Please review this issue on stackoverflow.

[Airflow "This DAG isnt available in the webserver DagBag object "](https://stackoverflow.com/questions/41560614/airflow-this-dag-isnt-available-in-the-webserver-dagbag-object)

Problem:

I have following code which dynamically creates Dag file and store it in /Users/[username]/airflow

```
#
# IMPORTS
#
import os  # CORE: Operating System module for file system operations.
import inspect  # CORE: Inspect module for getting source code of a function.
from datetime import datetime  # CORE: Datetime module for date and time operations.
from airflow.models import DAG, DagBag  # PIP: Airflow DAG and DagBag modules for DAG operations.
from airflow.operators.python_operator import PythonOperator  # PIP: Airflow PythonOperator module for Python operations.


#
# ENVS
#
os.environ['KMP_DUPLICATE_LIB_OK'] = 'True'


#
# FUNCTIONS
#
def DagFunctionToExecute():
    from datetime import datetime, timedelta
    from airflow import DAG
    from airflow.operators.bash_operator import BashOperator

    default_args = {
        'owner': 'airflow',
        'depends_on_past': False,
        'start_date': datetime(2023, 3, 16),
        'email': ['airflow@example.com'],
        'email_on_failure': False,
        'email_on_retry': False,
        'retries': 1,
        'retry_delay': timedelta(minutes=5),
    }

    dag = DAG('Amazing', default_args=default_args, schedule_interval=timedelta(days=1))

    t1 = BashOperator(
        task_id='print_date',
        bash_command='date',
        dag=dag)

    t2 = BashOperator(
        task_id='sleep',
        bash_command='sleep 5',
        retries=3,
        dag=dag)

    t1 >> t2


def CheckDagInDagBag(__id):
    # Load all tags from dagbag folder.
    dagbag = DagBag(dag_folder=os.path.join(os.environ['AIRFLOW_HOME'], 'dags'))

    print(dagbag.dags, os.path.join(os.environ['AIRFLOW_HOME'], 'dags'))

    # Check if DAG is present in the DagBag.
    if dagbag.dags.get(__id):
        # Print statement that dag is present.
        print(f"DAG with id {__id} is present in the DagBag.")
    else:
        # Print statement that dag is not present.
        print(f"DAG with id {__id} is not present in the DagBag.")


def CreateOperators(__dag, __edges, __functionToExecute):
    # Variable assignment.
    operators = {}

    # Loop over edges and convert them to operators.
    for item in __edges:
        source = item['source']
        target = item['target']

        # Check if source and target are not in operators.
        if source not in operators:
            # Update operator with source and python operator
            # with callable python function.
            operators[source] = PythonOperator(
                task_id=source,
                python_callable=__functionToExecute,
                provide_context=True,
                dag=__dag,
                op_kwargs={'source': source, '__dagId': __dag.dag_id}
            )

        # Check if target is not in operators.
        if target not in operators:
            # Update operator with target and python operator
            operators[target] = PythonOperator(
                task_id=target,
                python_callable=__functionToExecute,
                provide_context=True,
                dag=__dag,
                op_kwargs={'source': source, '__dagId': __dag.dag_id}
            )

        # Set upstream of target to source.
        operators[target].set_upstream(operators[source])

    # Return operators.
    return operators


def SyncDagToDb(__dag):
    # Sync DAG to database.
    __dag.sync_to_db()

    # Check if DAG is present in the DagBag.
    CheckDagInDagBag(__dag.dag_id)


def WriteDagToFile(__dag, __functionToExecute):
    # Variable assignment.
    dagId = __dag.dag_id
    fileName = f"{dagId}.py"

    # Get source code of DAG.
    sourceCode = inspect.getsource(__functionToExecute)

    # Write DAG to file.
    with open(os.path.join(os.environ['AIRFLOW_HOME'], 'dags', fileName), "w", encoding='utf8') as __dagFile:
        # Write source to the file.
        __dagFile.write(sourceCode)


def CreateDag(__edges, __title, __description):
    # Create DAG.
    dag = DAG(
        dag_id=__title,
        description=__description,
        start_date=datetime.now(),
        default_view='tree',
        schedule_interval='@daily',
        catchup=False
    )

    # Create operators and sync DAG to database.
    # along side to it write dag to file.
    CreateOperators(dag, __edges, DagFunctionToExecute)
    SyncDagToDb(dag)
    WriteDagToFile(dag, DagFunctionToExecute)

    # Return dag.
    return dag
```

Now when i Run `CreateDag` it creates file in dags folder and works like charm.

Screenshot of Dags Folder

<img width="918" alt="Screenshot 2023-03-17 at 3 25 49 AM" src="https://user-images.githubusercontent.com/76559517/225761133-e09b1497-7a77-423a-b622-f5faa3800938.png">

and Dag do appear in Airflow webApp screenshot is below.

<img width="1728" alt="Screenshot 2023-03-17 at 3 26 36 AM" src="https://user-images.githubusercontent.com/76559517/225761257-4d72228d-e0d9-4788-a28f-126c1f1d794c.png">

Problem is that when i click on Dag it throws error that **DAG "Amazinsg" seems to be missing from DagBag.** screenshot is below

<img width="1728" alt="Screenshot 2023-03-17 at 3 27 35 AM" src="https://user-images.githubusercontent.com/76559517/225761416-498edcee-d9ba-4288-a18e-c4dd82d104c4.png">

### How to reproduce

<img width="747" alt="Screenshot 2023-03-17 at 3 32 59 AM" src="https://user-images.githubusercontent.com/76559517/225762828-bc67c845-60bc-4a55-bb88-2628a74a25da.png">

### Operating System

OS Ventura

### Versions of Apache Airflow Providers

apache-airflow-providers-celery==3.1.0
apache-airflow-providers-common-sql==1.3.4
apache-airflow-providers-ftp==3.3.1
apache-airflow-providers-http==4.2.0
apache-airflow-providers-imap==3.1.1
apache-airflow-providers-sqlite==3.3.1

### Deployment

Virtualenv installation

### Anything else

💔 This issue is opened from last 6 Years on Stackoverflow and still it is unresolved. its a request to you guys please help me out in this i dont move away from airflow because i have already spent last 5 Days coding whole server around it. Hope you guys understand.

Stackoverflow: [https://stackoverflow.com/questions/41560614/airflow-this-dag-isnt-available-in-the-webserver-dagbag-object](Stackoverflow)

Regards

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

**Comment by boring-cyborg** [comment]
Thanks for opening your first issue here! Be sure to follow the issue template! If you are willing to raise PR to address this issue please do so, no need to wait for approval.

**Comment by potiuk** [comment]
This is not an issue with airflow, so I converted it back to discussion (the second one after you previously opened in https://github.com/apache/airflow/discussions/30148 

Again, let me explain it to you. You need to make sure that your DAG file can be properly parsed by scheduler. Check if that is happening - it might be that the generated source code is not a proper DAG and scheduler does not parse it. Do NOT write and sync DagBag to a database. This is an internal detail of Airflow (the db and how things are written to the DB is an internal detail of Airflow and might change any time without a warning, so you should absolutely not rely on it when trying to "interface" with Airflow.

There are multiple things that will stop working correctly if you "just" save dag to the DB. Scheduler perform way more things than that - this is only what DAGFileProcessor does (it can now even be run as separate, standalone component, but Scheduler must be running continuously, continue parsing the files and be responsible for making sure DAGs find its way from the DAG folder to the DB and that Airflow works.

The only "supported" way to create a new DAG in Airflow is to make the DAG file appears in the DAG folder and wait until scheduler parses and schedules it. Full stop. Any attempts to do it differently is a futile effort and even if you manage to make it works now it might stop working even in patchlevel version of Airflow, because we are free to change it (and we do - with pretty much every release). The API that your filesystem exposes to write the DAG file to the DAG folder is the "right" (and the only) one.

So - as explained before:

1) write your DAG file to DAG folder
2) have scheduler running
3) wait until scheduler processes it (make sure that your generated DAG file is parseable and recognized as DAG according to regular expectations of the DAG file) and it appears in the UI (you can use REST API to poll for the DAG to appar. 

Do not - under any circumstances - try to run external script that tries to reverse engineer whar dag file processor and scheduler does. This won't work. This will break for you continuously even if you manage to make it works for specific version.

You can also follow another path - you can lead and implement Airflow Improvement Proposal to enable similar behaviour. There were at least few attempts (none of them went past the Draft):

* https://cwiki.apache.org/confluence/display/AIRFLOW/AIP-5+Remote+DAG+Fetcher
* https://cwiki.apache.org/confluence/display/AIRFLOW/AIP-20+DAG+manifest

If you would like to follow this up, writing up the proposal/design for changes in Airflow Architecture to enable similar behaviour to what you expect is the first thing to do - theen discussing it at the devlist, leading to consensus that this is the right thing to do and finally implementing it is the right way to do it.
